src/materials/NEMLStress.C

```diff
@@ -21,17 +21,168 @@ InputParameters
 NEMLStress::validParams()
 {
   InputParameters params = NEMLStressBase::validParams();
+  params.addClassDescription("Use a constitutive model in the NEML library to compute stress");
   params.addRequiredParam<FileName>("database", "Path to NEML XML database.");
-  params.addRequiredParam<std::string>("model", "Model name in NEML database.");
+  params.addParam<std::vector<std::string>>("neml_variable_iname",
+                                            "Names of NEML xml name/value pairs");
+
+  params.addParam<Real>("neml_variable_value0",
+                        "NEML xml variable value for neml_variable_iname[0]");
+  params.addParam<Real>("neml_variable_value1",
+                        "NEML xml variable value for neml_variable_iname[1]");
+  params.addParam<Real>("neml_variable_value2",
+                        "NEML xml variable value for neml_variable_iname[2]");
+  params.addParam<Real>("neml_variable_value3",
+                        "NEML xml variable value for neml_variable_iname[3]");
+  params.addParam<Real>("neml_variable_value4",
+                        "NEML xml variable value for neml_variable_iname[4]");
+  params.addParam<Real>("neml_variable_value5",
+                        "NEML xml variable value for neml_variable_iname[5]");
+  params.addParam<Real>("neml_variable_value6",
+                        "NEML xml variable value for neml_variable_iname[6]");
+  params.addParam<Real>("neml_variable_value7",
+                        "NEML xml variable value for neml_variable_iname[7]");
+
+  params.addDeprecatedParam<std::string>(
+      "model",
+      "Model name in NEML database.",
+      "This parameter no longer has any effect. Simply remove it.");
+
   return params;
 }
 
-NEMLStress::NEMLStress(const InputParameters & parameters)
-  : NEMLStressBase(parameters),
-    _fname(getParam<FileName>("database")),
-    _mname(getParam<std::string>("model"))
+std::vector<Real> constructNemlSubstitutionList(const std::vector<std::string> & nemlNames);
+std::string parseAndReplaceXmlVariables(const std::vector<std::string> & nemlNames,
+                                        const std::vector<Real> & nemlValues,
+                                        const FileName & fname);
+
+NEMLStress::NEMLStress(const InputParameters & parameters) : NEMLStressBase(parameters)
+{
+  std::vector<std::string> nemlNames;
+  if (isParamValid("neml_variable_iname"))
+  {
+    nemlNames = getParam<std::vector<std::string>>("neml_variable_iname");
+  }
+
+  std::vector<Real> nemlValues = constructNemlSubstitutionList(nemlNames);
+
+  FileName fname(getParam<FileName>("database"));
+  std::string xmlStringForNeml = parseAndReplaceXmlVariables(nemlNames, nemlValues, fname);
+
+  _model = neml::parse_string_unique(xmlStringForNeml);
+}
+
+std::string
```

Comment:
Better to make these functions members of the class so they're not in the global namespace. When you're doing that, use doxygen to describe what these do and what the parameters are in the header.